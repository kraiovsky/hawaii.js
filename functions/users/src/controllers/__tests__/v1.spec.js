const { createMockContext } = require('@shopify/jest-koa-mocks')
const { createUser, findUser } = require('../v1')
const testCases = require('./v1.cases')

jest.mock('../../queries/users')

let ctx
beforeEach(() => {
  ctx = createMockContext()
  ctx = {
    ...ctx,
    set: jest.fn(headers => {
      ctx.headers = {
        ...ctx.headers,
        ...headers,
      }
    }),
    send: jest.fn((status, body) => {
      ctx.response = { status, body }
    }),
    ok: jest.fn(body => ctx.send(200, body)),
    throw: jest.fn(() =>
      ctx.send(500, {
        errors: [
          {
            status: 500,
            title: 'INTERNAL_SERVER_ERROR',
            detail: 'Internal Server Error',
          },
        ],
      })
    ),
    fail: jest.fn(({ code, msg, info }) =>
      ctx.send(code || 500, {
        errors: [
          {
            status: code || 500,
            title: msg || 'INTERNAL_SERVER_ERROR',
            detail: info || 'ctx.fail called',
          },
        ],
      })
    ),
  }
})
afterEach(() => {
  jest.clearAllMocks()
})

describe('createUser()', () => {
  describe.each(testCases.createUser.happy)('👍', (should, email) => {
    test(`should ${should}`, async () => {
      ctx.request = {
        body: {
          email,
        },
      }
      await createUser()(ctx)
      expect(ctx.send).toHaveBeenCalled()
      expect(ctx.response).toMatchSnapshot()
    })
  })
})

describe('findUser()', () => {
  describe.each(testCases.findUser.happy)('👍', (should, fields, email) => {
    test(`should ${should}`, async () => {
      ctx.query = {
        'fields[users]': fields,
        searchKey: 'email',
      }
      ctx.params = {
        searchQuery: email,
      }
      await findUser()(ctx)
      expect(ctx.ok).toHaveBeenCalled()
      expect(ctx.response).toMatchSnapshot()
    })
  })
})
