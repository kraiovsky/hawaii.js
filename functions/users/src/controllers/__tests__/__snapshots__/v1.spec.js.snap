// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createUser() üëç should return existing user 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "avatar": "http://lorempixel.com/800/800/avatar?66700",
        "email": "email@domain.com",
        "name": "Test",
        "role": "user",
      },
      "id": "0ae76c84-d939-456d-bb35-f60b7659ed6b",
      "links": Object {
        "self": "/users/0ae76c84-d939-456d-bb35-f60b7659ed6b",
      },
      "relationships": undefined,
      "type": "user",
    },
    "included": undefined,
    "jsonapi": Object {
      "version": "1.0",
    },
    "links": Object {
      "index": "/users",
    },
    "meta": undefined,
  },
  "status": 200,
}
`;

exports[`createUser() üëç should return new user 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "avatar": null,
        "email": "newemail@domain.com",
        "name": null,
        "role": "user",
      },
      "id": "e9ad2f8f-b871-4048-95ff-29577f4285f5",
      "links": Object {
        "self": "/users/e9ad2f8f-b871-4048-95ff-29577f4285f5",
      },
      "relationships": undefined,
      "type": "user",
    },
    "included": undefined,
    "jsonapi": Object {
      "version": "1.0",
    },
    "links": Object {
      "index": "/users",
    },
    "meta": undefined,
  },
  "status": 201,
}
`;

exports[`findUser() üëç should return user with all non-blacklisted if empty field provided in query 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "avatar": "http://lorempixel.com/800/800/avatar?66700",
        "email": "email@domain.com",
        "name": "Test",
        "role": "user",
      },
      "id": "0ae76c84-d939-456d-bb35-f60b7659ed6b",
      "links": Object {
        "self": "/users/0ae76c84-d939-456d-bb35-f60b7659ed6b",
      },
      "relationships": undefined,
      "type": "user",
    },
    "included": undefined,
    "jsonapi": Object {
      "version": "1.0",
    },
    "links": Object {
      "index": "/users",
    },
    "meta": undefined,
  },
  "status": 200,
}
`;

exports[`findUser() üëç should return user with all non-blacklisted if no field provided in query 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "avatar": "http://lorempixel.com/800/800/avatar?66700",
        "email": "email@domain.com",
        "name": "Test",
        "role": "user",
      },
      "id": "0ae76c84-d939-456d-bb35-f60b7659ed6b",
      "links": Object {
        "self": "/users/0ae76c84-d939-456d-bb35-f60b7659ed6b",
      },
      "relationships": undefined,
      "type": "user",
    },
    "included": undefined,
    "jsonapi": Object {
      "version": "1.0",
    },
    "links": Object {
      "index": "/users",
    },
    "meta": undefined,
  },
  "status": 200,
}
`;

exports[`findUser() üëç should return user with email and name 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "email": "email@domain.com",
        "name": "Test",
      },
      "id": "0ae76c84-d939-456d-bb35-f60b7659ed6b",
      "links": Object {
        "self": "/users/0ae76c84-d939-456d-bb35-f60b7659ed6b",
      },
      "relationships": undefined,
      "type": "user",
    },
    "included": undefined,
    "jsonapi": Object {
      "version": "1.0",
    },
    "links": Object {
      "index": "/users",
    },
    "meta": undefined,
  },
  "status": 200,
}
`;
